src/vs/workbench/contrib/notebook/browser/notebookKernelServiceImpl.ts

```diff
@@ -96,8 +138,33 @@ export class NotebookKernelService extends Disposable implements INotebookKernel
 		}
 	}
 
+	private _initSourceActions() {
+		const loadActionsFromMenu = (menu: IMenu) => {
+			const groups = menu.getActions({ shouldForwardArgs: true });
+			const actions: IAction[] = [];
+			groups.forEach(group => {
+				actions.push(...group[1]);
+			});
+			this._sourceActions = actions.map(action => {
+				const sourceAction = new SourceAction(action);
+				const stateChangeListener = sourceAction.onDidChangeState(() => {
+					this._onDidChangeSourceActions.fire();
+				});
+				return [sourceAction, stateChangeListener];
+			});
+			this._onDidChangeSourceActions.fire();
+		};
+
+		this._register(this._sourceMenu.onDidChange(() => {
+			loadActionsFromMenu(this._sourceMenu);
+		}));
+
+		loadActionsFromMenu(this._sourceMenu);
+	}
+
 	override dispose() {
 		this._kernels.clear();
+		this._sourceActions.forEach(a => a[1].dispose());
```

Comment:
nit: you can use the `dispose`-utility for this, e.g `dispose(this._sourceActions.values())`